### PR TITLE
reduce k8s-example image by 104MB

### PR DESCRIPTION
**Pull request by alex-treebeard**
Experiment with @mrdavidlaing using  [whaler](https://github.com/treebeardtech/whaler) to remove unnecessary image contents.

An additive approach may be nicer than this, using [distroless](https://github.com/GoogleContainerTools/distroless) as a base for the output image.

### CODE REVIEW

**Comment by mrdavidlaing** [comment]
@alex-treebeard Thanks for this great start.

I'm struggling to reproduce the 104MB savings; and wondering if I'm doing something wrong.

```
$ docker images | grep k8s-example
dagster/k8s-example                                             latest                  b89711f17093   4 weeks ago      436MB
k8s-example-minimal                                             latest                  f8b3a87102d3   14 minutes ago   427MB
```

I'm building [`dagster/python_modules/automation/automation/docker/images/k8s-example/Dockerfile`](https://github.com/dagster-io/dagster/blob/76a13a55a7f33d6fc92f9a2ae0ff4123a9497414/python_modules/automation/automation/docker/images/k8s-example/Dockerfile) using the following build command:

```
 docker build --build-arg BASE_IMAGE=python:3.7-slim --build-arg DAGSTER_VERSION=0.9.22.post0 --tag=k8s-example-minimal -f dagster/python_modules/automation/automation/docker/images/k8s-example/Dockerfile .
```

Which `--build-arg` are you using so we can compare apples to apples?

**Comment by mrdavidlaing** [comment]
@alex-treebeard Could you possibly give me write access to your fork so I can also update this PR?

I'd like to add this:

```
# (54 MB) Remove all cached Python files
RUN find / -regex '^.*\(__pycache__\|\.py[co]\)$' -delete
```

**Comment by alex-treebeard** [comment]
Gave you write access to this fork!

I made the mistake of only looking at the size of the last layer. Starting from scratch ensures there is only one layer solving this problem.

This is probably bad practice! Feel free to alter as you wish.

**Comment by alex-treebeard** [comment]
By the way: Will deleting these pyc files/caches have any other impact on the image. E.g. container start times?

**Comment by mrdavidlaing** [comment]
@alex-treebeard Thanks for write access!

Re: deleting pyc files/caches -> slower start times - possibly; although I'm unsure.  Its quite a large size saving; so I guess it depends whether the reduction network / disk cache load is worth the increased CPU load/time at startup.

Let leave that out for now, since I want to discuss and alternative to having a single layer on top of scratch that I think might be better from a caching perspective.

Poking around the original image layers with [`dive`](https://github.com/wagoodman/dive) shows this:

image

I think its likely that the first set of layers will already be cached at the deploy target:

 - the first 69MB layer - `FROM 174542ec711eff4` - is from `debian:buster-slim`
 - the next 4 layers (42MB) are from `python:3.7.8-slim`
 
 Layers 6 & 7 are added by Dagster; and are likely to be cachable across ALL the user code images deployed by a user
 
 I propose that focusing on squashing & minimising the "Dagster code" layers 6 & 7 might be optimal in terms of getting the smallest image that takes maximum advantage of caching.
 
 ie: 
image

What do you think to switching to this approach as shown in the commit below?

**Comment by alex-treebeard** [comment]
Thanks a lot for the write up, your commit is the right way forward.

I had a little look at tools that automate this type of optimisation you have done on the final layer and found [minicon](https://github.com/grycap/minicon) which seems interesting. Perhaps there is a more succinct way of packaging up this space saving for Python apps?

In any case I may write a short blog on this case study, let me know if you have any interesting output metrics e.g. on deployment times off the back of this, or follow-up ideas!

**Comment by mrdavidlaing** [comment]
[minicon](https://github.com/grycap/minicon) looks fascinating!  Have you tried it on our use-case?

I'd be happy to help contribute some metrics for the blog post; although anecdotally I haven't noticed a dramatic difference :(

Proposal for next steps - lets try get the "optimised layers" approach above merged into master; and then start a separate exploration using minicon

File: python_modules/automation/automation/docker/images/k8s-example/Dockerfile [hunk 1]
```diff
@@ -1,34 +1,27 @@
...
+    && apt-get install -yqq cron \
+# Dagster
+    && pip install \
+        dagster==${DAGSTER_VERSION} \
+        dagster-postgres==${DAGSTER_VERSION} \
+        dagster-celery[flower,redis,kubernetes]==${DAGSTER_VERSION} \
+        dagster-aws==${DAGSTER_VERSION} \
+        dagster-k8s==${DAGSTER_VERSION} \
+        dagster-celery-k8s==${DAGSTER_VERSION} \
+# Cleanup
```
Comment (mrdavidlaing) [inline python_modules/automation/automation/docker/images/k8s-example/Dockerfile:19]: Paging @chenbobby - would you be amenable to merging  a change like this ( that minimised the "dagster" layer ) in all the standard docker images ?

**Review by chenbobby — APPROVED** [review]
This is a fantastic optimization. thanks for working on this, @alex-treebeard and @mrdavidlaing 🙏 ⭐ 

i would also be excited to see if `whaler` and `minicon` reveal oppportunities for minimizing [the other Docker images in `automation/docker/image`](https://github.com/dagster-io/dagster/tree/master/python_modules/automation/automation/docker/images) 😁

**Pull request merged**